File: .claude/hooks/post-test-analysis.php
#!/usr/bin/env php
<?php
/**
 * PostToolUse hook: Suggest Codex analysis after test/build failures.
 *
 * Analyzes test and build output and suggests Codex consultation
 * for debugging complex failures.
 */

// Commands that run tests or builds
const TEST_BUILD_COMMANDS = [
    'phpunit',
    'pest',
    'composer test',
    'npm test',
    'npm run test',
    'npm run build',
    'phpstan',
    'php-cs-fixer',
    'psalm',
    'tsc',
    'cargo test',
    'go test',
    'make test',
    'make build',
];

// Patterns indicating failures that need debugging
const FAILURE_PATTERNS = [
    '/FAILED/',
    '/ERROR/',
    '/error\\[/',
    '/Error:/',
    '/failed/',
    '/error:/',
    '/Fatal error/',
    '/TypeError/',
    '/ValueError/',
    '/Exception/',
    '/Stack trace/',
    '/PHPUnit\\\\Framework\\\\/',
    '/FAILURES!/',
    '/Tests:.*failures/',
    '/panic:/',
    '/FAIL:/',
];

// Simple errors that don't need Codex
const SIMPLE_ERRORS = [
    'Class not found',
    'command not found',
    'No such file or directory',
];

function isTestOrBuildCommand(string $command): bool
{
    $commandLower = strtolower($command);
    foreach (TEST_BUILD_COMMANDS as $cmd) {
        if (str_contains($commandLower, $cmd)) {
            return true;
        }
    }
    return false;
}

/**
 * Check if output contains complex failures that need debugging.
 *
 * @return array{bool, string}
 */
function hasComplexFailure(string $output): array
{
    // Skip if it's a simple error
    foreach (SIMPLE_ERRORS as $simple) {
        if (str_contains($output, $simple)) {
            return [false, ''];
        }
    }

    // Count failure patterns
    $failureCount = 0;
    foreach (FAILURE_PATTERNS as $pattern) {
        $matches = [];
        if (preg_match_all($pattern . 'i', $output, $matches)) {
            $failureCount += count($matches[0]);
        }
    }

    // Multiple failures or complex errors suggest need for Codex
    if ($failureCount >= 3) {
        return [true, "Multiple failures detected ({$failureCount} issues)"];
    }

    // Single failure in test output
    $outputLower = strtolower($output);
    if ($failureCount >= 1 && (str_contains($outputLower, 'stack trace') || str_contains($outputLower, 'assertion'))) {
        return [true, 'Test failure with stack trace'];
    }

    return [false, ''];
}

function main(): void
{
    try {
        $input = file_get_contents('php://stdin');
        $data = json_decode($input, true);
        $toolName = $data['tool_name'] ?? '';

        // Only process Bash tool
        if ($toolName !== 'Bash') {
            exit(0);
        }

        $toolInput = $data['tool_input'] ?? [];
        $toolOutput = $data['tool_output'] ?? '';
        $command = $toolInput['command'] ?? '';

        // Check if it's a test/build command
        if (!isTestOrBuildCommand($command)) {
            exit(0);
        }

        // Check for complex failures
        [$hasFailure, $reason] = hasComplexFailure($toolOutput);

        if ($hasFailure) {
            $output = [
                'hookSpecificOutput' => [
                    'hookEventName' => 'PostToolUse',
                    'additionalContext' =>
                        "[Codex Debug Suggestion] {$reason}. "
                        . 'Consider consulting Codex for debugging analysis. '
                        . "**Recommended**: Use Task tool with subagent_type='general-purpose' "
                        . 'to consult Codex with full error context and preserve main context.',
                ],
            ];
            echo json_encode($output) . "\n";
        }

        exit(0);

    } catch (\Throwable $e) {
        fwrite(STDERR, "Hook error: {$e->getMessage()}\n");
        exit(0);
    }
}

main();

File: .claude/hooks/post-test-analysis.py
#!/usr/bin/env python3
"""
PostToolUse hook: Suggest Codex analysis after test/build failures.

Analyzes test and build output and suggests Codex consultation
for debugging complex failures.
"""

import json
import re
import sys

# Commands that run tests or builds
TEST_BUILD_COMMANDS = [
    "pytest",
    "npm test",
    "npm run test",
    "npm run build",
    "uv run pytest",
    "ruff check",
    "ty check",
    "mypy",
    "tsc",
    "cargo test",
    "go test",
    "make test",
    "make build",
]

# Patterns indicating failures that need debugging.
#
# Every pattern is anchored or quantified so that it can only match a runner's
# *verdict*, never a test name, a file name, or prose. Bare tokens ("ERROR",
# "failed", "Error:") matched a green `pytest -v` run whose test names merely
# contained the word "error", so the hint fired on 8 passed, 0 failed.
# Matched case-sensitively for the same reason: "error" in an identifier is not
# an error, "ERROR" at the start of a line is.
FAILURE_PATTERNS = [
    # Runner verdict lines: pytest "FAILED test_x", "ERROR test_x".
    r"(?m)^(?:FAILED|ERROR)\b",
    # pytest's failure-detail prefix ("E   assert 1 == 2").
    r"(?m)^E\s",
    # Summary counts: "1 failed, 2 passed".
    r"\b\d+\s+failed\b",
    # An exception header at the start of a line, qualified or not:
    # "AssertionError: ...", "django.db.Error: ...", "Exception: ...".
    r"(?m)^\s*(?:\w+\.)*\w*(?:Error|Exception):\s",
    # rustc / cargo diagnostics.
    r"(?m)^error\[\w+\]",
    # Go panics and gotest/ctest verdict lines.
    r"(?m)^panic:",
    r"(?m)^FAIL\b",
    # Python traceback header (parenthesised, so it cannot match a test name).
    r"Traceback \(most recent call last\):",
    # Named exception types as whole words; case-sensitive, so the lowercase
    # forms that appear inside identifiers do not match.
    r"\b(?:AssertionError|TypeError|ValueError|AttributeError|ImportError|"
    r"ModuleNotFoundError|SyntaxError)\b",
]

# Failure count at which the output is reported as a multi-failure run.
MIN_FAILURES_FOR_MULTIPLE = 3

# Simple errors that don't need Codex
SIMPLE_ERRORS = [
    "ModuleNotFoundError",  # Usually just need to install
    "command not found",
    "No such file or directory",
]


def is_test_or_build_command(command: str) -> bool:
    """Check if the command runs tests or builds."""
    command_lower = command.lower()
    return any(cmd in command_lower for cmd in TEST_BUILD_COMMANDS)


def has_complex_failure(output: str) -> tuple[bool, str]:
    """Check if output contains complex failures that need debugging."""
    # Skip if it's a simple error
    for simple in SIMPLE_ERRORS:
        if simple in output:
            return False, ""

    # Count failure patterns
    failure_count = 0
    matched_patterns = []
    for pattern in FAILURE_PATTERNS:
        matches = re.findall(pattern, output)
        if matches:
            failure_count += len(matches)
            matched_patterns.append(pattern)

    # Any failure in a test or build command is worth a Codex look: a red
    # suite is the cheapest possible moment to find the root cause, and a
    # single failure is not evidence that the cause is simple.
    if failure_count >= MIN_FAILURES_FOR_MULTIPLE:
        return True, f"Multiple failures detected ({failure_count} issues)"

    if failure_count >= 1:
        plural = "" if failure_count == 1 else "s"
        if any(
            p in output.lower()
            for p in ["traceback", "assertion", "error", "exception"]
        ):
            return (
                True,
                f"Test failure with error details ({failure_count} issue{plural})",
            )
        return True, f"Test/build failure detected ({failure_count} issue{plural})"

    return False, ""


def build_context(data: dict) -> str | None:
    """Return the additionalContext hint for this hook, or None.

    Pure function (no stdin/stdout/exit) so the post-bash-check.py
    dispatcher can call it in-process. Standalone main() below still
    reads stdin directly for backwards-compatible direct invocation.
    """
    tool_name = data.get("tool_name", "")
    # Only process Bash tool
    if tool_name != "Bash":
        return None

    tool_input = data.get("tool_input", {})
    tool_response = data.get("tool_response", {})
    command = tool_input.get("command", "")
    tool_output = tool_response.get("stdout", "") or tool_response.get("content", "")

    # Check if it's a test/build command
    if not is_test_or_build_command(command):
        return None

    # Check for complex failures
    has_failure, reason = has_complex_failure(tool_output)
    if not has_failure:
        return None

    return (
        f"[Codex Debug Suggestion] {reason}. "
        "Use the `codex-debugger` subagent before attempting a manual fix — "
        "deep reasoning finds root causes that surface-level fixes miss. "
        "**Recommended**: Task(subagent_type='codex-debugger') with the full "
        "command and test output, which also preserves main context."
    )


def main():
    try:
        data = json.load(sys.stdin)
        context = build_context(data)

        if context:
            output = {
                "hookSpecificOutput": {
                    "hookEventName": "PostToolUse",
                    "additionalContext": context,
                }
            }
            print(json.dumps(output))

        sys.exit(0)

    except Exception as e:
        print(f"Hook error: {e}", file=sys.stderr)
        sys.exit(0)


if __name__ == "__main__":
    main()
